### PR TITLE
Improve documentation of `compute()` function

### LINKED ISSUE #116697: `compute` function different from describe?
``` dart
import 'dart:isolate';

import 'package:flutter/foundation.dart';

class BackgroundPlus {
  final PlusData plusData;

  BackgroundPlus(this.plusData);

  Future<int> computeInBackground() async {
    return compute(_computeEntry, plusData);
  }

  int _computeEntry(PlusData plusData) {
    print(
        "_computeEntry isolate: ${Isolate.current.debugName}@${Isolate.current.hashCode}");
    return plusData.a + plusData.b;
  }
}

class PlusData {
  final int a;
  final int b;

  PlusData(this.a, this.b);
}

void main() async {
  print(
      "main isolate: ${Isolate.current.debugName}@${Isolate.current.hashCode}");
  final data = await BackgroundPlus(PlusData(2, 3)).computeInBackground();
  print("data=$data");
}

```

After I run code below, this message on console is: 

image

Which means I can run `compute` function with memeber function and user-defined class. But the descibe of the `compute` function is:

- `callback`:

``` dart

/// {@template flutter.foundation.compute.callback}
/// Qualifying functions include:
///
///   * top-level functions
///   * static methods
///   * closures that only capture objects that can be sent to an isolate

```

- `message`

``` dart

/// There are limitations on the values that can be sent and received to and
/// from isolates. These limitations constrain the values of `Q` and `R` that
/// are possible. See the discussion at [SendPort.send].

```

Why this works? One possible reason I guess is that `Isolate.spawn` will copy all memory hold by its `message` param to the new isolate. After I read descibe of `Isolate.spaw` carefully, I didn't find any limitation on `message` param. So I guess the `_IsolateConfiguration` params passed to `compute` function will be copied too. Since callback is a member funtion pointer, the  whole object will be deep copied too. 

If my guess is correct, the describe of `compute` is misleading user force use limited function and message type; Or else, use member function and user defined class member will bring some harm?

### PR DESCRIPTION
**Pull request by mkustermann**
This improves the documentation of the `compute()` function as follows:

  * Instead of making `compute` a top-level constant, we make it a function. This allows the generated API docs to show a function signature with parameters and their names, making it *much* clearer to users what function is being documented.

  * We mention that on web-backends this is running `compute()` on the normal eventloop whereas on on native platforms it runs in a separate isolate.

  * We mention that callback, message and result have to be sendable across isolates. We also mention that they may be copied.

  * We link to both `Isolate.run` & `SendPort.send` for more information.

  * We use `M` for the type of `message` instead the rather confusing `Q`.

Closes https://github.com/flutter/flutter/issues/116697

### CODE REVIEW

File: packages/flutter/lib/src/foundation/isolates.dart [hunk 1]
```diff
@@ -16,26 +16,15 @@ import '_isolates_io.dart'
...
-/// {@macro flutter.foundation.compute.usecase}
-///
-/// The function used as `callback` must be one that can be sent to an isolate.
-/// {@macro flutter.foundation.compute.callback}
-///
-/// {@macro flutter.foundation.compute.types}
-///
-/// The `debugLabel` argument can be specified to provide a name to add to the
-/// [Timeline]. This is useful when profiling an application.
 typedef ComputeImpl = Future<R> Function<Q, R>(ComputeCallback<Q, R> callback, Q message, { String? debugLabel });
```
Comment (aam) [inline packages/flutter/lib/src/foundation/isolates.dart:24]: Q->M here as well?

File: packages/flutter/lib/src/foundation/isolates.dart [hunk 2]
```diff
@@ -16,26 +16,15 @@ import '_isolates_io.dart'
 /// {@macro flutter.foundation.compute.callback}
```
Comment (aam) [inline packages/flutter/lib/src/foundation/isolates.dart:16]: I believe this reference and the one below point to deleted macros.

Comment (mkustermann) [reply packages/flutter/lib/src/foundation/isolates.dart:12]: Thanks. It makes now sense to me, a `@macro` is basically copying in code from the `@template`.
Done.

**Comment by mkustermann** [comment]
@aam PTAL

**Review by aam — APPROVED** [review]

**Pull request merged**